datalad/tests/fixtures.py

```diff
@@ -0,0 +1,73 @@
+"""utilities to assist with setting up tests"""
+
+from datalad.api import clone
+from datalad.core.distributed.clone import decode_source_spec
+from datalad.distribution.dataset import Dataset
+from datalad.interface.common_cfg import dirs
+from datalad.utils import Path, better_wraps, optional_args
+from datalad.tests.utils import with_tempfile
+
+
+DATALAD_TESTS_CACHE = Path(dirs.user_cache_dir) / "tests"
+
+
+def get_cached_dataset(url, dataset_name=None, version=None):
+    """ Helper to get a cached clone from url
+
+    Intended for use from within `datalad_dataset` decorator.
+    Clones `url` into user's cache under datalad/tests/`name`. If such a clone
+    already exists, don't clone but return the existing one. So, it's supposed
+    to cache the original source in order to reduce time and traffic for tests,
+    by letting subsequent requests clone from a local location directly.
+
+    If it's an annex get the content, too.
+
+    Note
+    ----
+    `version` not yet implemented. Intention: verify that `version` can be
+    checked out, but don't actually do it, since the cached dataset is intended
+    to be used as origin instead of the original remote at URL by the
+    `datalad_dataset` test decorator. Checkout of a particular should happen
+
+    Parameters
+    ----------
+    url: str
+        URL to clone from
+    dataset_name: str
+        (directory) name to use for the clone
+
+    Returns
+    -------
+    Dataset
+    """
+
+    # TODO: What about recursive?
+
+    # TODO: a (env var) switch to use tmp location rather than user's cache?
+    #       would be: persistent across tests vs persistent across test runs.
+
+    if not dataset_name:
+        dataset_name = decode_source_spec(url)['default_destpath']
+
+    ds = Dataset(DATALAD_TESTS_CACHE / dataset_name)
+    if not ds.is_installed():
+        ds = clone(url, ds.pathobj)
+    # TODO: check version
+    ds.get()
+    return ds
+
+
+@optional_args
+def datalad_dataset(f, url=None, version=None, name=None):
+
+    # TODO: env var to enable/disable caching? Or melt with location switch
+    #       in get_cached_dataset?
+
+    @better_wraps(f)
+    @with_tempfile
+    def newfunc(*arg, **kw):
+        ds = get_cached_dataset(url, name, version=version)
+        clone_ds = clone(ds.pathobj, arg[-1])
```

Comment:
In many cases clone wouldn't be needed, option? That is why I thought to make get also optional in the original cached one

Reply:
May be also could be reckless? Pass clone kwargs?

Reply:
I think it should always be a clone. We shouldn't actually operate on datasets that are used across tests. And it wouldn't be a cache, if we deliver it directly. Getting a cached dataset shouldn't have any side effects compared to getting it directly from that URL (which would give you - a clone of the original one). To me the entire point is to be equivalent to cloning from that URL to a temp. location from the point of view of a test. Except that it becomes cheaper across tests and test runs in terms of network traffic and time needed. Cache serves just as a drop-in replacement for the original URL. And with the suggested env var it could then be disabled and clone directly from that URL (or use a temp. cache instead of user's). Therefore env var makes no difference to the test at all, but only to its performance.

Reply:
Re clone kwargs:
Guess that's useful, although I start being afraid of having to write tests for all those features and their combinations.

Reply:
Thinking about it a bit more:
May be it should be two decorators (don't want to much features in one thing):

1. as described above: replaces the remote one specified by `url` by a cached one as the origin and delivers a clone thereof. If caching is disabled, delivers a direct clone of the original remote.

2. Doesn't deliver a clone, but a URL to be used in the test. original URL if caching is disabled and path to a cached dataset (or `file:` scheme URL) if enabled.

What do you think about this, @yarikoptic ?

Reply:
Done. `@cached_dataset` + `@cached_url` instead of `@datalad_dataset`.